### ISSUE TITLE
Spring webflow: redirect url contains String[] parameter values

### ISSUE THREAD

**Issue by aukevanleeuwen** [Jasig/cas#1593]
If I specify the attribute `redirect="true"` on a view in one of the webflows (`login-webflow.xml` for example) I end up with serialized `String[]` in the URL. So the URL would look like:

`/login?client_name=%5BLjava.lang.String%3B%40221ebc09&oauth_token=%5BLjava.lang.String%3B%40fcb045f&oauth_verifier=%5BLjava.lang.String%3B%40598c0112`

url-decoded:

`/login?client_name=[Ljava.lang.String;@221ebc09&oauth_token=[Ljava.lang.String;@fcb045f&oauth_verifier=[Ljava.lang.String;@598c0112`

I'm pretty sure the culprit is in the the following class: `org.jasig.cas.web.flow.CasDefaultFlowUrlHandler` in the method:

```
    @Override
    public String createFlowExecutionUrl(final String flowId, final String flowExecutionKey, final HttpServletRequest request) {
        final StringBuilder builder = new StringBuilder();
        builder.append(request.getRequestURI());
        builder.append('?');
        @SuppressWarnings("unchecked")
        final Map<String, Object> flowParams = new LinkedHashMap<String, Object>(request.getParameterMap());
        flowParams.put(this.flowExecutionKeyParameter, flowExecutionKey);
        appendQueryParameters(builder, flowParams, getEncodingScheme(request));
        return builder.toString();
    }
```

Here `request.getParameterMap()` will return a `Map<String, String[]>` which results in the url above. Apparently `org.springframework.webflow.context.servlet.DefaultFlowUrlHandler#appendQueryParameters` wasn't meant to be fed a `Map<String, String[]>` (which I find weird, since you exactly expect it to receive the parameter map from the request). But this doesn't work really well.

**Comment by mmoayyed** [comment]
Thanks for the report. Possible ideas on the fix?

**Comment by mmoayyed** [comment]
Should we explode that array as individual parameters?

**Comment by mmoayyed** [comment]
Quick workaround: could u just do an externalRedirect like some of the other views?

**Comment by aukevanleeuwen** [comment]
Yeah the proper way is to explode the array to individual parameters. Basically if you have a request with querystring: `?key1=value1.1&key1=value1.2&key2=value2.1` calling r`equest.getParameterMap()` will return (pseudo-json):

```
{
  key1: [value1.1, value1.2]
  key2: [value2.1]
}
```

and this should go the other way around.

**Comment by mmoayyed** [comment]
Fair enough. Could I interest you to taking this one and put together a few confirming tests in a PR?

**Comment by aukevanleeuwen** [comment]
I'd like to have this end up in the 4.2.x, should I PR against 4.2.x or master? (i.e. do you forward-port or back-port)?

I'm pretty pressed for time, but maybe tomorrow evening? I'll see what I can do.

**Comment by mmoayyed** [comment]
You can do either. We'll port it where necessary. I'd go for master. 

Note that we are about 10 days away from the GA release. So you should plenty of time. 

Thanks again.

**Comment by aukevanleeuwen** [comment]
I can't seem to build cas, any idea what I'm doing wrong? I don't have a lot of gradle experience...

```
cas 16h master / ./gradlew build --parallel -x test -x javadoc -DskipCheckstyle=true -DskipAspectJ=true -DskipFindbugs=true                                                                                                                                             java-1.8.0_65
Parallel execution is an incubating feature.

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/auke/projects/workspaces/nrc/cas/build.gradle' line: 1017

* What went wrong:
A problem occurred evaluating root project 'cas-server'.
> No such property: java for class: org.gradle.api.java.archives.internal.DefaultManifest

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 5.516 secs
```

**Comment by mmoayyed** [comment]
Are you building from master? Are you recent in your fork? Make sure you're using the latest JDK 1.8.

**Comment by aukevanleeuwen** [comment]
Version information:

```
------------------------------------------------------------
Gradle 2.10
------------------------------------------------------------

Build time:   2015-12-21 21:15:04 UTC
Build number: none
Revision:     276bdcded730f53aa8c11b479986aafa58e124a6

Groovy:       2.4.4
Ant:          Apache Ant(TM) version 1.9.3 compiled on December 23 2013
JVM:          1.8.0_65 (Oracle Corporation 25.65-b01)
OS:           Mac OS X 10.10.5 x86_64
```
Apparently I needed to add

```
apply plugin: 'java'
```

in `build.gradle` for it to work. But even after that I got the following error:


```
uild file '/Users/auke/projects/workspaces/nrc/cas/build.gradle' line: 1286

* What went wrong:
A problem occurred evaluating root project 'cas-server'.
> Cannot add task ':javadoc' as a task with that name already exists.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 10.933 secs
```

I commented the :javadoc task out for now, after that I build succesfully. This was a fully up to date clone of Jasig/cas master.

**Comment by aukevanleeuwen** [comment]
This might be a bit confusing, but I created two pull requests for both `master` and `4.2.x` because I later on realised that `4.2.x` was on Java 7 of course. So I rewrote it without the lambda's.

**Comment by mmoayyed** [comment]
You did perfectly great. Thanks.

**Comment by aukevanleeuwen** [comment]
Sorry about the styling issues, trying to fix those now. takes a while to get used to all the finals ;-)

**Comment by aukevanleeuwen** [comment]
Oh btw for a bit more information about the *why* I use this feature: I'm using the pac4j integration to login with a couple of social networks. However upon returning from a successful validation I'm trying to find a linked account for this social network (i.e. I have a single account that can be linked to multiple social networks). I have extended the webflow in case no such account is found one is automatigically created or linked to an existing account.

The pac4j integration goes something like this:
* redirect to social network (save some stuff in the session)
* login
* come back to callback url (retrieve state from session)

now the callback url looks something like this (top of my head): `/login?client=someClientId&state=some-difficult-state-identifier&even=more&param=eters`. Now at that point I extended the webflow with some extra screens, but the URL at that point never changes. For one that is a bit ugly, but more than that: If the user presses refresh at some point it will usually error out. It's been a while since I implemented this, so I'm a bit fuzzy about the details, but I think either on the CAS side some session infrormation has been removed _or_ the social network does not accept the `accessToken` anymore to retrieve the `UserProfile` again. Either way there was something with kind of a one-time state that would trigger `ClientAction` to fail.

Now that I use `redirect=true`, my URL changes and the user can press reload.
